lib/kubernetes-deploy/kubernetes_resource.rb

```diff
@@ -54,19 +41,28 @@ def timeout
       self.class.timeout
     end
 
-    def initialize(name:, namespace:, context:, file:, logger:)
+    def initialize(namespace:, context:, template:, logger:)
       # subclasses must also set these if they define their own initializer
-      @name = name
+      @name = template.fetch("metadata", {})["name"]
       @namespace = namespace
       @context = context
-      @file = file
       @logger = logger
+      @template = template
     end
 
     def id
       "#{type}/#{name}"
     end
 
+    def file
```

Comment:
This can also result to tempfiles not being garbage collected after the process is done...
If we're already using it this way today then it doesn't have to be addressed in this PR - but may be something that we want to revisit in future. WDYT?